File: Lecture24/relax.py
from math import *
from bvp import *

print " Relaxing solution of u'' = - pi^2 (u+1) / 4"
print " with boundary conditions u(0) = 0, u(1) = 1";
acc = input(" Enter desired accuracy: ")

print ""
print " Iteration          Change              Error"
print " -----------------------------------------------------"

relaxer = BVPRelax( x0=0., x1=1., u0=0., u1=1., N=100 )

relaxer.guess()
relaxer.relax()
iteration = 1
while relaxer.error() > acc:
    relaxer.relax();
    iteration += 1
    print " %6d      %18.16g  %18.16g" % (iteration, relaxer.change(), relaxer.error())

file = open("relaxing.data", "w")
for i in range(relaxer.N+1):
    x = relaxer.x0 + i * relaxer.dx
    file.write(repr(x) + "\t" + repr(relaxer.u[i]) + "\n")
file.close()
print " N =", relaxer.N, "step solution in file relaxing.data"

File: Lecture24/relax.cpp
#include "cptstd.hpp"
#include "matrix.hpp"
using namespace cpt;
#include "bvp.hpp"

int main() {
    cout << " Relaxing solution of u'' = - pi^2 (u+1) / 4\n"
         << " with boundary conditions u(0) = 0, u(1) = 1\n";
    cout << " Enter desired accuracy: ";
    double acc;
    cin >> acc;

    cout << "\n Iteration          Change              Error"
         << "\n -----------------------------------------------------\n";

    BVPRelax relax;
    relax.initial_guess();
    relax.Jacobi_step();
    int iteration = 1;
    relax.print(iteration);
    while (relax.error() > acc) {
      relax.Jacobi_step();
      relax.print(++iteration);
    }

    ofstream data_file("relaxing.data");
    for (int i = 0; i <= relax.get_N(); i++)
      data_file << relax.get_u_lb() + i * relax.get_h() << '\t' << relax.get_u(i) << '\n';
    data_file.close();
    cout << "\n " << iteration << " step solution in file relax.data"
         << endl;
}
